src/main/java/org/craftercms/core/cache/impl/store/guava/GuavaCacheStoreAdapter.java

```diff
@@ -0,0 +1,159 @@
+ * it under the terms of the GNU General Public License version 3 as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.craftercms.core.cache.impl.store.guava;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.craftercms.core.cache.CacheItem;
+import org.craftercms.core.cache.CacheStatistics;
+import org.craftercms.core.cache.impl.CacheStoreAdapter;
+import org.springframework.beans.factory.DisposableBean;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Implementation of a {@link CacheStoreAdapter} using Guava {@link Cache}s.
+ *
+ * @author joseross
+ */
+public class GuavaCacheStoreAdapter implements CacheStoreAdapter, DisposableBean {
+
+    protected final Map<String, Cache<Object, Object>> caches = new ConcurrentHashMap<>();
+
+    /**
+     * Destroy method, called by the Spring container. Calls {@link Cache#cleanUp()} for all instances.
+     */
+    @Override
+    public void destroy() {
+        caches.forEach((scope, cache) -> {
+            cache.invalidateAll();
+            cache.cleanUp();
+        });
+    }
+
+    @Override
+    public boolean hasScope(String scope) {
+        return caches.containsKey(scope);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Collection<String> getScopes() {
+        return caches.keySet();
+    }
+
+    /**
+     * Adds a new scope. The scope is an instance of Guava's {@link Cache}.
+     *
+     * @param scope            the name of the scope
+     * @param maxItemsInMemory the maximum number of items in memory, before they are evicted
+     * @throws Exception
+     */
+    @Override
+    public void addScope(String scope, int maxItemsInMemory) {
+        caches.put(scope, CacheBuilder.newBuilder().recordStats().maximumSize(maxItemsInMemory).build());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void removeScope(String scope) {
+        caches.remove(scope);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getSize(String scope) {
+        return (int) caches.get(scope).size();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Collection<Object> getKeys(String scope) {
+        return caches.get(scope).asMap().keySet();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasKey(String scope, Object key) {
+        return caches.get(scope).asMap().containsKey(key);
```

Comment:
Does this create a new map every time or does this just return the underlying data structure? Because if it returns a new map every time this might not be very performant.

Reply:
It returns the internal instance https://github.com/google/guava/blob/v29.0/guava/src/com/google/common/cache/LocalCache.java#L4919